test/test_helpers/thread_helper.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module ThreadHelper
+  # This allows some part of the code to run in another Thread
+  # The issue is that Rails allocate a new connection on the new Thread
+  # So the global wrapping transaction is obviously not seen by this new Thread
+  # When Rails rolls back the global transaction, this thread has already
+  # committed changes to the DB.
+  def within_thread(*args)
```

Comment:
Another alternative is to set the connection like this

```ruby
connection = ActiveRecord::Base.connection
Thread.new do
  ActiveRecord::Base.connection = connection
  yield
end.join
```

This is done in Rails 5.1 https://github.com/rails/rails/commit/deba47799ff905f778e0c98a015789a1327d5087
Waiting for upgrade before doing it